components/causal_ts/src/hlc.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use pd_client::PdClient;
+use std::{
+    fmt,
+    sync::{
+        atomic::{AtomicBool, AtomicU64, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+use tikv_util::worker::{Builder as WorkerBuilder, Worker};
+use txn_types::TimeStamp;
+
+use crate::{
+    errors::{Error, Result},
+    CausalTsProvider,
+};
+
+/// An implementation of Hybrid Logical Clock (HLC)
+/// See proposal for detail:
+/// https://github.com/tikv/rfcs/blob/master/text/0083-rawkv-cross-cluster-replication.md
+#[derive(Debug)]
+struct Hlc(AtomicU64);
+
+impl Default for Hlc {
+    fn default() -> Self {
+        Self(AtomicU64::new(0))
+    }
+}
+
+impl fmt::Display for Hlc {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0.load(Ordering::Relaxed))
+    }
+}
+
+impl Hlc {
+    #[inline]
+    fn next(&self) -> TimeStamp {
+        let ts = self.0.fetch_add(1, Ordering::Release).into();
+        trace!("Hlc::next: {:?}", ts);
+        ts
+    }
+
+    // self.next() == `to`
+    #[inline]
+    fn advance(&self, to: TimeStamp) {
+        let before = self.0.fetch_max(to.into_inner(), Ordering::Acquire).into();
+        debug!("Hlc::advance"; "before" => ?before, "after" => ?std::cmp::max(before, to));
+    }
+}
+
+const TSO_REFRESH_INTERVAL: Duration = Duration::from_millis(500);
+
+/// A causal timestamp provider by HLC
+pub struct HlcProvider {
+    hlc: Arc<Hlc>,
+    pd_client: Arc<dyn PdClient>,
+    tso_worker: Worker,
+
+    /// Interval of Physical Clock refresh from TSO.
+    /// Smaller interval will get more accurate metrics, but more pressure on PD.
+    /// Can be Duration::ZERO, which means never refresh from TSO. For test purpose ONLY.
+    tso_refresh_interval: Duration,
+
+    initialized: AtomicBool,
```

Comment:
I suggest initializing HLC in `HlcProvider::new_opt` so that `HlcProvider` must be initialized.

Reply:
OK. The `init()` is an async method, as it invoking `pd_client.get_tso()`. Then I will use `block_on(pd_client.get_tso())` in `new_opt()`.